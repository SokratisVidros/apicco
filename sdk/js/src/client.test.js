/* eslint-env jest */

const nock = require('nock');
const apiccoClient = require('./client');

nock.disableNetConnect();

describe('Apicco Client', () => {
  let apicco;

  beforeEach(async () => {
    nock('http://apicco.test')
      .get('/api/v1/discovery')
      .reply(200, {
        'fruits.list': [],
        'fruits.info': ['*fruit_id'],
        'fruits.eat': ['*name', '*vitamins', 'ripe']
      });

    apicco = await apiccoClient({
      origin: 'http://apicco.test',
      relPath: 'api/v1'
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('discovers all API methods along with their parameters automatically', () => {
    expect(apicco).toMatchSnapshot();
  });

  describe('request method on success', () => {
    beforeEach(() => {
      nock('http://apicco.test')
        .post('/api/v1/fruits.list')
        .reply(200, ['🍉', '🍊', '🍋']);
    });

    it('exposes a try catch api', async () => {
      const fruits = await apicco.fruits.list();
      expect(fruits).toEqual(['🍉', '🍊', '🍋']);
    });

    it('exposes a promisified api', async () => {
      await expect(apicco.fruits.list()).resolves.toEqual(['🍉', '🍊', '🍋']);
    });

    it('request methods exposes a callback api', (done) => {
      apicco.fruits.list((err, fruits) => {
        expect(err).toBeNull();
        expect(fruits).toEqual(['🍉', '🍊', '🍋']);
        done();
      });
    });
  });

  describe('request method on error', () => {
    beforeEach(() => {
      nock('http://apicco.test')
        .post('/api/v1/fruits.list', '{"bar":"foo"}')
        .reply(400, {
          error: 'Bad Request',
          message: 'Invalid Request Body',
          statusCode: 400
        });
    });

    it('exposes an async await promisified api', async () => {
      try {
        await apicco.fruits.list({ bar: 'foo' });
      } catch (err) {
        expect(err).toEqual({
          error: 'Bad Request',
          message: 'Invalid Request Body',
          statusCode: 400
        });
      }
    });

    it('exposes a thenable promisified api', async () => {
      await expect(apicco.fruits.list({ bar: 'foo' })).rejects.toEqual({
        error: 'Bad Request',
        message: 'Invalid Request Body',
        statusCode: 400
      });
    });

    it('exposes a callback api', (done) => {
      apicco.fruits.list({ bar: 'foo' }, (err, fruits) => {
        expect(err).toEqual({
          error: 'Bad Request',
          message: 'Invalid Request Body',
          statusCode: 400
        });
        expect(fruits).toBeUndefined();
        done();
      });
    });
  });

  describe('request method when response is empty', () => {
    beforeEach(() => {
      nock('http://apicco.test')
        .post('/api/v1/fruits.list')
        .reply(null);
    });

    it('exposes a promisified api', async () => {
      const fruits = await apicco.fruits.list();
      expect(fruits).toEqual('');
    });
  });

  describe('validations', () => {
    beforeEach(() => {
      nock('http://apicco.test')
        .post('/api/v1/fruits.eat')
        .reply(204);
    });

    it('validate required parameters for the action request', async () => {
      try {
        await apicco.fruits.eat();
      } catch (err) {
        expect(err.message).toBe('fruits.eat missing params: name, vitamins');
      }
    });
  });
});
