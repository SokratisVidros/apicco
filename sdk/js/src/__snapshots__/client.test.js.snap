// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apicco Client discovers all API methods along with their parameters automatically 1`] = `
Object {
  "fruits": Object {
    "eat": [Function],
    "info": [Function],
    "list": [Function],
  },
}
`;

exports[`Apicco Client request method with credential headers send requests using credential:same-origin (default) when no Access-Control headers are set on server 1`] = `
Array [
  Array [
    "http://apicco.test/api/v1/fruits.list",
    Object {
      "body": undefined,
      "credentials": "same-origin",
      "headers": Object {
        "Content-Type": "application/json",
        "User-Agent": "Apicco JS SDK/1.0.10",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`Apicco Client request method with credential headers sends requests using credential:include when Access-Control-Allow-Credentials is set on server 1`] = `
Array [
  Array [
    "http://apicco.test/api/v1/fruits.list",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Content-Type": "application/json",
        "User-Agent": "Apicco JS SDK/1.0.10",
      },
      "method": "POST",
    },
  ],
]
`;
